### PR TITLE
Tdl 21520 skip donotprocess records

### PR DESCRIPTION
**Pull request by RushiT0122**
# Description of change
If any visitor is set the Do Not Process flag then Pendo will stop collecting events from that visitor and will stop displaying guides to that visitor. In such scenario we should skip fetching visitor_history.

# Manual QA steps
 - On cloned connection verified that `donotprocess` records are getting skipped.
 - Updated 'donotprocess' of few visitors in the test account to cover this scenario in the integration tests
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_pendo/streams.py [hunk 1]
```diff
@@ -378,8 +377,16 @@ def sync_substream(self, state, parent, sub_stream, parent_response):
...
+                        # If any visitor is set the Do Not Process flag then Pendo will stop collecting events
+                        # from that visitor and will stop displaying guides to that visitor
+                        LOGGER.info("Record marked as 'Do Not Process': %s", record[parent.key_properties[0]])
+                        continue
+
                     last_processed = self.get_last_processed(state, sub_stream)
-                    parent_last_updated = datetime.fromtimestamp(float(record['metadata']['auto'][self.replication_key]) / 1000.0, timezone.utc)
+                    # If there is no last_update key then set 'last_updated=now()' and fetch the records from recent bookmark
+                    parent_last_updated = datetime.fromtimestamp(float(
+                        record['metadata']['auto'][self.replication_key]) / 1000.0, timezone.utc) if record['metadata']['auto'].get(self.replication_key) else now()
```
Comment (rdeshmukh15) [inline tap_pendo/streams.py:389]: 140 characters on one line which is too long. Can you please align it properly. 100 chars per line are preferred.

File: .circleci/config.yml [hunk 2]
```diff
@@ -22,7 +22,7 @@ jobs:
           command: |
             source /usr/local/share/virtualenvs/tap-pendo/bin/activate
             # TODO: Adjust the pylint disables
-            pylint tap_pendo --disable 'broad-except,chained-comparison,empty-docstring,fixme,invalid-name,line-too-long,missing-class-docstring,missing-function-docstring,missing-module-docstring,no-else-raise,no-else-return,too-few-public-methods,too-many-arguments,too-many-branches,too-many-lines,too-many-locals,ungrouped-imports,wrong-spelling-in-comment,wrong-spelling-in-docstring,bad-whitespace,missing-class-docstring'
+            pylint tap_pendo --disable 'broad-except,chained-comparison,empty-docstring,fixme,invalid-name,line-too-long,missing-class-docstring,missing-function-docstring,missing-module-docstring,no-else-raise,no-else-return,too-few-public-methods,too-many-arguments,too-many-branches,too-many-lines,too-many-locals,ungrouped-imports,wrong-spelling-in-comment,wrong-spelling-in-docstring,bad-whitespace,missing-class-docstring,too-many-statements'
```
Comment (rdeshmukh15) [inline .circleci/config.yml:25]: break into multiple lines for readability purpose.

File: tap_pendo/streams.py (hunk 1 above)
Comment (RushiT0122) [reply tap_pendo/streams.py]: Done

File: tap_pendo/streams.py [hunk 3]
```diff
@@ -378,8 +377,16 @@ def sync_substream(self, state, parent, sub_stream, parent_response):
 
                 # Filtering the visitors based on last updated to improve performance of visitor_history replication
                 if isinstance(parent, Visitors):
+                    if record['metadata'].get('pendo', {'donotprocess': False}).get('donotprocess'):
+                        # If any visitor is set the Do Not Process flag then Pendo will stop collecting events
+                        # from that visitor and will stop displaying guides to that visitor
```
Comment (dsprayberry) [inline tap_pendo/streams.py:382]: Can we update this docstring to explain why we skip those records, eg they will not contain required keys and we can't persist them?

**Review by dsprayberry — CHANGES_REQUESTED** [review]

**Review by dsprayberry — APPROVED** [review]

File: tap_pendo/streams.py (hunk 1 above)
Comment (dsprayberry) [reply tap_pendo/streams.py]: Do we know under what scenarios a record will not have a last_update key? Is it linked exclusively to visitors with do_not_process flags set?

**Pull request merged**